Add support for propeller outputs urls

igorsobreira opened the pull request:
Propeller added support for multiple outputs (https://github.com/cbsinteractive/propeller/pull/895)

Right now bakery support the following URLs:
```
/propeller/orgID/channelID.m3u8
/propeller/orgID/clip/clipID.m3u8
```
this change will add support for
```
/propeller/orgID/channelID/outputID.m3u8
```

Required change on propeller-go client: https://github.com/cbsinteractive/propeller-go/pull/4

**Note on tests**

The first commit in this PR adds some tests for `origin.Configure()` 
for propeller endpoints. The test starts a fake Propeller API with stubbed
responses allowing origin.NewPropeller() to interact with the API
and build a `origin.Propeller` object

For situations like this I usually prefer to mock the go client itself,
but Bajery used uses a reference to `*propeller.Client` which makes it
harder to replace for tests. We could refactor that into an interface,
if Bakery works with a `propeller.Client` interface then it's easy to swap
with a mock implementation for tests.

I didn't want to do any design changes before adding support for
propeller outputs urls. So this test helps increase coverage for
`origin.NewPropeller()`

zsiec commented on origin/origin_test.go:
```diff
@@ -217,3 +249,58 @@ func TestOrigin_Configure(t *testing.T) {
+//
+// Make sure to call teardown() when the test is over
+func configMockPropellerAPI() (cfg config.Config, teardown func()) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
👍 on using interfaces to avoid the need for a construction like this

igorsobreira replied:
Which interface are you referring to?

zsiec replied:
I don't think the interface exists yet, I was just agreeing that a client interface (once it exists) will help simplify here :). We do this with [most](https://github.com/cbsinteractive/drm-svc/blob/master/client/drmsvc/client.go#L18) of our [clients](https://github.com/cbsinteractive/transcode-orchestrator/blob/master/client/transcoding/client.go#L12)

zsiec replied:
I see you added them later on in this PR 💯

zsiec commented on origin/origin_test.go:
```diff
@@ -217,3 +249,58 @@ func TestOrigin_Configure(t *testing.T) {
+		}
+	}))
+
+	tsURL, _ := url.Parse(ts.URL)
```
even in tests, I think it's prudent to check errors and fail the test

igorsobreira replied:
I didn't check because an error there would mean Go's `httptest.NewServer()` returned an invalid url, which is very unlikely.
5bc757d3957ee403b46afa8e4454147417f13270

zsiec replied:
I agree, quite unlikely :) I think it sets a good precedent though. In this case, rather than `panic` we could supply a `*testing.T` to this method and simply call `t.Fail` which would be less surprising than a panic during testing.

igorsobreira replied:
8b9cf2b

zsiec commented on origin/propeller.go:
```diff
@@ -18,16 +19,16 @@ type Propeller struct {
-
-	if len(parts) < 4 {
-		err := fmt.Errorf("propeller origin: request format is not `/propeller/orgID/channelID.m3u8`")
+	urlValues, err := parsePropellerPath(path)
```
Unrelated directly to these changes, it may make sense to eventually be more explicit here perhaps adopting functionality similar to http.ServeMux if we expect this to grow at all in the future.

igorsobreira replied:
Yes, that would be the next step. Right now it's easy to identify which one matched based on the fields provided so I didn't want to add unnecessary complexity

zsiec commented on origin/propeller.go:
```diff
@@ -36,13 +37,14 @@ func configurePropeller(c config.Config, path string) (Origin, error) {
-	orgID := parts[2]
+	orgID := urlValues["orgID"]
+	channelID := urlValues["channelID"]
+	clipID := urlValues["clipID"]
```
one suggestion would be to hoist these values into constants used here and in the regex capture group definition to avoid them diverging accidentally

igorsobreira replied:
I'm not sure I understand the suggestion. What do you mean by "hoist these values into constants [...]"?

zsiec replied:
`"orgID:`, `"channelID"` and `"clipID"` could be defined as constants used in the regex definitions and in these map lookups

zsiec commented on origin/propeller.go:
```diff
@@ -76,6 +78,27 @@ func NewPropeller(c config.Config, orgID string, endpointID string, get fetchURL
+		}
+		return values, nil
+	}
+	return map[string]string{}, errors.New("propeller origin: request format is not `/propeller/orgID/channelID.m3u8`")
```
Another benefit of being more explicit: this error message seems to mention one specific format when more than one is really supported and that could cause confusion to a consumer.

Also, is there any reason to create and return a map here on error? Why not just `nil`?

igorsobreira replied:
Good catch, I updated the error message: f761310b916301a9ca40911735d6be75c91d93e2

I used to always return `nil` when an error is being returned, but I noticed [blank values being returned with errors](https://github.com/cbsinteractive/bakery/blob/d7cdc02492e64204d5a9f46537962ea94901d8ad/origin/propeller.go#L71) instead of `nil` so I tried to follow the standard. But since you asked then it's probably not a standard, so I changed to `nil` :)

zsiec replied:
Gotcha, yeah makes sense to look at project context! In this case, I think we can advocate for always returning the "zero" value (nil if pointer, slice, map, channel etc, empty struct if not a pointer) here and in https://github.com/cbsinteractive/bakery/blob/d7cdc02492e64204d5a9f46537962ea94901d8ad/origin/propeller.go#L71 and everywhere else :)

igorsobreira replied:
Yep, agree

krystalmejia24 commented on origin/origin_test.go:
```diff
@@ -217,3 +249,58 @@ func TestOrigin_Configure(t *testing.T) {
+// API that returns hard-coded responses
+//
+// Make sure to call teardown() when the test is over
+func configMockPropellerAPI() (cfg config.Config, teardown func()) {
```
I do really love these sets of tests you had written here. Initially, my idea was eventually to mock the client itself as you said had stated in your notes 

> If Bakery works with a propeller.Client interface then it's easy to swap
with a mock implementation for tests.

If I had more time (lazy excuse), this was the direction I was planning on going. When I first created the origin package, propeller-go client, I didn't know enough at the time to do that. I did encourage this approach for goofy so @sam-eng  can write her tests much easier. If you want to create the interface and refactor the tests, I'd be all for it! But this is a great way to ensure the end to end functionality when configuring an origin channel.

At some point I considered putting this mocked client under a `test/` package in the propeller-go client: https://github.com/cbsinteractive/goofy/blob/schedule-api/schedule/server/handlers_test.go#L17
so that all of our projects that interface with the client can just import it but im not sure if i was being overzealous with that approach.

igorsobreira replied:
I like to have some of those higher level tests to make sure all objects are being used correctly then test all the small details with unit tests.

My bad for not seeing the unit tests earlier on `origin/propeller_test.go`.

So the approach I take for these unit tests is I specify an interface with a subset of methods used by the propeller client:
https://github.com/cbsinteractive/bakery/blob/0b5ae30442065c737409ceb57af7016da053e869/origin/propeller.go#L111-L115

and I use that inside my module
https://github.com/cbsinteractive/bakery/blob/0b5ae30442065c737409ceb57af7016da053e869/origin/propeller.go#L120
this makes it easy to replace for a mock client.

But I think a better approach would be for the client itself to give us an interface, either `propeller.Client` is an interface (and the client is a `DefaultClient` type) or, if the `propeller-go` only exposes a client the interface could be `propeller.Interface`. This way Bakery's mock can implement that interface and doesn't need the `propellerClient` I created.

If `propeller-go` gives us an interface then it could also provide a Mock implementation of that client like you suggested.

krystalmejia24 replied:
Yea my thought process here was heading towards `propeller-go` supplying us an interface so that it could provide those mock implementations since we have at least two code bases using this. 

And no worries on missing those earlier tests. These changes make the code base so much cleaner and easier to read. lets get rid of my elementary code!

igorsobreira commented:
Summary of changes:

**Refactor URL dispatcher for Propeller origin**

The propeller origin accepts multiple url formats, and with a new format being added I felt that the [current approach of `strings.Split(path, "/")`](https://github.com/cbsinteractive/bakery/blob/d7cdc02492e64204d5a9f46537962ea94901d8ad/origin/propeller.go#L27) was getting hader to read.

So I added a [list of patterns we accept using regex](https://github.com/cbsinteractive/bakery/blob/0b5ae30442065c737409ceb57af7016da053e869/origin/propeller.go#L15-L19) and a [function `parsePropellerPath`](https://github.com/cbsinteractive/bakery/blob/0b5ae30442065c737409ceb57af7016da053e869/origin/propeller.go#L90-L109) to perform the match and extract relevant fields from the URL.

It's still very simple. For instance, we only know [which URL matched based on the fields returned](https://github.com/cbsinteractive/bakery/blob/0b5ae30442065c737409ceb57af7016da053e869/origin/propeller.go#L43-L55). For now this looked enough but a second step would be to give a "name" for each URL and return it so the caller knows which one, or dispatch directly to a method associated to that pattern. Doesn't feel like we need this complexity yet.

**Refactor `fetchURL` functions. Add new one for Channel Output**

`origin/propeller.go` [had a `fetchURL` function type](https://github.com/cbsinteractive/bakery/blob/d7cdc02492e64204d5a9f46537962ea94901d8ad/origin/propeller.go#L19) to allow both implementations [`getPropellerChannelURL`](https://github.com/cbsinteractive/bakery/blob/d7cdc02492e64204d5a9f46537962ea94901d8ad/origin/propeller.go#L79-L110) and [`getPropellerClipURL`](https://github.com/cbsinteractive/bakery/blob/master/origin/propeller.go#L112-L134) to be used by [`NewPropeller`](https://github.com/cbsinteractive/bakery/blob/d7cdc02492e64204d5a9f46537962ea94901d8ad/origin/propeller.go#L59-L62).

I changed that to [an interface `urlGetter`](https://github.com/cbsinteractive/bakery/blob/0b5ae30442065c737409ceb57af7016da053e869/origin/propeller.go#L119-L121). Now we have [`channelURLGetter`](https://github.com/cbsinteractive/bakery/blob/0b5ae30442065c737409ceb57af7016da053e869/origin/propeller.go#L123-L154) and [`clipURLGetter`](https://github.com/cbsinteractive/bakery/blob/0b5ae30442065c737409ceb57af7016da053e869/origin/propeller.go#L207-L233). And, the purpose of the PR, a new [`outputURLGetter`](https://github.com/cbsinteractive/bakery/blob/0b5ae30442065c737409ceb57af7016da053e869/origin/propeller.go#L156-L189)

**origin.Configure() functional tests**

You guys already saw the tests I added using a mocked Propeller API server. After working on `origin/propeller.go` I noticed there were unit tests for it. So I updated the `origin/origin_test.go` tests to [go over only the happy path for Channel, Clip, and Output cases](https://github.com/cbsinteractive/bakery/blob/0b5ae30442065c737409ceb57af7016da053e869/origin/origin_test.go#L174-L191). 

For those tests we mock the Propeller API itself and call `origin.Configure()`. Those are important to have at least once for each propeller entity because they check if we are handling the different URLs correctly

igorsobreira commented:
Question: how to solve this problem on the CI?
```
Run go test --race -v -coverprofile=coverage.out ./...
verifying github.com/cbsinteractive/propeller-go@v0.0.0-20200503222720-e53e98ec6b80/go.mod: github.com/cbsinteractive/propeller-go@v0.0.0-20200503222720-e53e98ec6b80/go.mod: reading https://sum.golang.org/lookup/github.com/cbsinteractive/propeller-go@v0.0.0-20200503222720-e53e98ec6b80: 410 Gone
```

I had the same problem locally and setting `GOPRIVATE` solved it
```
$ go env |grep -i private
GOPRIVATE="github.com/cbsinteractive/propeller-go"
```

zsiec commented:
> Question: how to solve this problem on the CI?
> 
> ```
> Run go test --race -v -coverprofile=coverage.out ./...
> verifying github.com/cbsinteractive/propeller-go@v0.0.0-20200503222720-e53e98ec6b80/go.mod: github.com/cbsinteractive/propeller-go@v0.0.0-20200503222720-e53e98ec6b80/go.mod: reading https://sum.golang.org/lookup/github.com/cbsinteractive/propeller-go@v0.0.0-20200503222720-e53e98ec6b80: 410 Gone
> ```
> 
> I had the same problem locally and setting `GOPRIVATE` solved it
> 
> ```
> $ go env |grep -i private
> GOPRIVATE="github.com/cbsinteractive/propeller-go"
> ```

I think we've been using `go env -w "GONOSUMDB=github.com/cbsinteractive"` in CI for this kind of thing. https://github.com/cbsinteractive/mediahub/blob/master/Dockerfile#L13

zsiec commented on origin/propeller_test.go:
```diff
@@ -1,250 +1,378 @@
-			} else if err == nil && tc.expectError {
-				t.Errorf("getClipURL() expected error, got nil")
+			u, err := tc.getter.GetURL(tc.client)
+			if err != nil && tc.expectErrStr == "" {
+				t.Errorf("returned unexpected error: %q", err)
+			} else if err == nil && tc.expectErrStr != "" {
+				t.Errorf("expected error, got nil")
+			} else if err != nil && err.Error() != tc.expectErrStr {
+				t.Errorf("got wrong error. expected: %q, got %q", tc.expectErrStr, err.Error())
 			}
```
If we expect an error, do we want to continue with the test here? In other projects we've use a construct like:

```go
if err != nil {
	if g, e := err.Error(), tt.wantErr; g != e {
		t.Errorf("Func() wrong error returned, got: %v, want: %v", g, e)
	}
	return
} else if tt.wantErr != "" {
	t.Error("Func() expected an error, got nil")
	return
}

```

igorsobreira replied:
No strong personal preferences here. In this case, I kept going because the returned URL can give a hint on why the error did not match. And I know the return value will not be nil (it's an empty string)

But as a general rule, I agree, it's dangerous to try and use the value is the error doesn't match

zsiec replied:
:+1: We should probably update `t.Errorf("expected error, got nil")` to use `t.Error` if we're not passing args to format.

igorsobreira replied:
Yep, good catch 96fd547
